src/IO/PeekableWriteBuffer.cpp

```diff
@@ -0,0 +1,85 @@
+#include <IO/PeekableWriteBuffer.h>
+
+namespace DB
+{
+
+PeekableWriteBuffer::PeekableWriteBuffer(DB::WriteBuffer & sub_buf_) : BufferWithOwnMemory(0), sub_buf(sub_buf_)
+{
+    Buffer & sub_working = sub_buf.buffer();
+    BufferBase::set(sub_working.begin() + sub_buf.offset(), sub_working.size() - sub_buf.offset(), 0);
+}
+
+void PeekableWriteBuffer::nextImpl()
+{
+    if (checkpoint)
```

Comment:
In what case is it useful to call this function when checkpoint is not set?

Reply:
What do you mean? `nextImpl()` method is the default interface of `ReadBuffer`, we cannot avoid calling it. If we already created `PeekableWriteBuffer` over some buffer, we will always use this method to write data, we cannot use underlying buffer, otherwise next time we want to use `PeekableWriteBuffer` it's state can be invalid.

Reply:
> I meant more like adding an assert (checkpoint)
Whoever is using PeekableWriteBuffer should set checkpoint before next is called right ?

Actually no, we can create `PeekableWriteBuffer` in advance and use checkpoints only when we need it. In formats we do like that because we should create read buffer in constructor and pass it to the base classes, it will be painful to create `PeekableWriteBuffer` on each `consume/write` call when we need checkpoints and propogate this buffer to all base classes